Several tests are unstable due to "CRITICAL: could not ingest record" error raised on pageserver shutdown

The following failures of miscellaneous tests during the past 3 days:
https://neon-github-public-dev.s3.amazonaws.com/reports/pr-10986/14198382944/index.html#testresult/ec386964bbbe7054
test_pageserver_catchup_while_compute_down[release-pg17] /  __arch: 'ARM64'
E   AssertionError: First log error on pageserver_1: (148, '2025-04-01T15:07:49.884156Z ERROR wal_connection_manager{tenant_id=5a9a7690f4508ffceafb4dbf8c992a2c shard_id=0000 timeline_id=26fd6defae3a6fd5339a8cfc21cb54b7}:connection{node_id=1}: CRITICAL: could not ingest record at 0/1503478\n')

https://neon-github-public-dev.s3.amazonaws.com/reports/pr-11382/14193402982/index.html#testresult/213f4174ca6483f5
test_crafted_wal_end[release-pg17-vanilla-last_wal_record_crossing_segment]  /  __arch: 'X64' 
E   AssertionError: First log error on pageserver_1: (314, '2025-04-01T10:46:40.376277Z ERROR wal_connection_manager{tenant_id=4a7be8fbcff3ea5a9ce31ecaf1a9e290 shard_id=0000 timeline_id=4c9ee8151e973027a956739fd4cc27a9}:connection{node_id=1}: CRITICAL: could not ingest record at 0/254D760\n')

https://neon-github-public-dev.s3.amazonaws.com/reports/main/14128409990/index.html#testresult/e99c27ee8c1f3a2c
test_pageserver_restarts_under_worload[release-pg17] / __arch: 'X64' 
E   AssertionError: First log error on pageserver_1: (1166, '2025-03-28T12:16:33.043077Z ERROR wal_connection_manager{tenant_id=2e15899d8b44c7e4f49398c1917b871b shard_id=0000 timeline_id=91f6d150bd3866f9b420ba8a7b4b22be}:connection{node_id=1}: CRITICAL: could not ingest record at 0/B01BBA0\n')

```
test.log:
2025-03-28 12:16:31.682 INFO [neon_fixtures.py:2320] Status call returned policy='Active' availability='WarmingUp' (193 attempts left)
2025-03-28 12:16:32.784 INFO [neon_cli.py:68] Running command "/tmp/neon/bin/neon_local pageserver stop --id=1"
2025-03-28 12:16:33.092 INFO [neon_cli.py:68] Running command "/tmp/neon/bin/neon_local pageserver start --id=1"
2025-03-28 12:16:33.206 INFO [neon_fixtures.py:2306] Polling 1 for None scheduling policy and Active availability
# the above message was produced after "pageserver stop" and before "pageserver start"


pageserver.log:
2025-03-28T12:16:32.790769Z  INFO wal_connection_manager{tenant_id=2e15899d8b44c7e4f49398c1917b871b shard_id=0000 timeline_id=91f6d150bd3866f9b420ba8a7b4b22be}:connection{node_id=1}:tokio_epoll_uring_ext::thread_local_system{thread_local=7 attempt_no=0}: successfully launched system
2025-03-28T12:16:32.791302Z  INFO Got signal SIGTERM. Terminating gracefully in fast shutdown mode.
2025-03-28T12:16:32.791324Z  INFO wal_connection_manager{tenant_id=2e15899d8b44c7e4f49398c1917b871b shard_id=0000 timeline_id=91f6d150bd3866f9b420ba8a7b4b22be}:connection{node_id=1}: caught up at LSN 0/B01BAE0
...
2025-03-28T12:16:32.791494Z  INFO shutdown_pageserver{exit_code=0}: completed stage="shutdown LibpqEndpointListener" elapsed_ms=0
2025-03-28T12:16:32.791539Z  INFO shutdown_pageserver{exit_code=0}:shutdown: Waiting for 0 InProgress tenants and 1 Attached tenants to shut down
2025-03-28T12:16:32.791647Z  INFO shutdown_pageserver{exit_code=0}:shutdown:shutdown{tenant_id=2e15899d8b44c7e4f49398c1917b871b shard_id=0000}: Waiting for timelines...
...
2025-03-28T12:16:33.043077Z ERROR wal_connection_manager{tenant_id=2e15899d8b44c7e4f49398c1917b871b shard_id=0000 timeline_id=91f6d150bd3866f9b420ba8a7b4b22be}:connection{node_id=1}: CRITICAL: could not ingest record at 0/B01BBA0

Caused by:
    timeline shutting down

Stack backtrace:
   0: anyhow::error::<impl core::convert::From<E> for anyhow::Error>::from
             at /__w/neon/neon/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/anyhow-1.0.94/src/backtrace.rs:27:14
   1: <core::result::Result<T,F> as core::ops::try_trait::FromResidual<core::result::Result<core::convert::Infallible,E>>>::from_residual
             at /rustc/4d91de4e48198da2e33413efdcd9cd2cc0c46688/library/core/src/result.rs:2014:27
   2: pageserver::pgdatadir_mapping::DatadirModification::ingest_batch::{{closure}}
             at /__w/neon/neon/pageserver/src/pgdatadir_mapping.rs:1604:31
   3: pageserver::walingest::WalIngest::ingest_record::{{closure}}
             at /__w/neon/neon/pageserver/src/walingest.rs:273:14
   4: pageserver::tenant::timeline::walreceiver::walreceiver_connection::handle_walreceiver_connection::{{closure}}
             at /__w/neon/neon/pageserver/src/tenant/timeline/walreceiver/walreceiver_connection.rs:441:26
   5: pageserver::tenant::timeline::walreceiver::connection_manager::ConnectionManagerState::change_connection::{{closure}}::{{closure}}::{{closure}}
             at /__w/neon/neon/pageserver/src/tenant/timeline/walreceiver/connection_manager.rs:563:18
...
   0: pageserver::tenant::timeline::walreceiver::walreceiver_connection::handle_walreceiver_connection::{{closure}}::{{closure}}
             at /__w/neon/neon/pageserver/src/tenant/timeline/walreceiver/walreceiver_connection.rs:449:33
   1: core::result::Result<T,E>::inspect_err
             at /rustc/4d91de4e48198da2e33413efdcd9cd2cc0c46688/library/core/src/result.rs:901:13
   2: pageserver::tenant::timeline::walreceiver::walreceiver_connection::handle_walreceiver_connection::{{closure}}
             at /__w/neon/neon/pageserver/src/tenant/timeline/walreceiver/walreceiver_connection.rs:439:21
   3: pageserver::tenant::timeline::walreceiver::connection_manager::ConnectionManagerState::change_connection::{{closure}}::{{closure}}::{{closure}}
             at /__w/neon/neon/pageserver/src/tenant/timeline/walreceiver/connection_manager.rs:563:18
   4: <tracing::instrument::Instrumented<T> as core::future::future::Future>::poll
             at /__w/neon/neon/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tracing-0.1.41/src/instrument.rs:321:9
   5: pageserver::tenant::timeline::walreceiver::TaskHandle<E>::spawn::{{closure}}
             at /__w/neon/neon/pageserver/src/tenant/timeline/walreceiver.rs:181:53
   6: <core::pin::Pin<P> as core::future::future::Future>::poll
             at /rustc/4d91de4e48198da2e33413efdcd9cd2cc0c46688/library/core/src/future/future.rs:124:9

```

with the seemingly critical error "could not ingest record" indicate a defect in the pageserver shutdown procedure.

Namely, Tenant::shutdown() performs:
```
...
        match self.set_stopping(shutdown_progress, false, false).await {
...
        let mut js = tokio::task::JoinSet::new();
        {
            let timelines = self.timelines.lock().unwrap();
            timelines.values().for_each(|timeline| {
                let timeline = Arc::clone(timeline);
                let timeline_id = timeline.timeline_id;
                let span = tracing::info_span!("timeline_shutdown", %timeline_id, ?shutdown_mode);
                js.spawn(async move { timeline.shutdown(shutdown_mode).instrument(span).await });
            });
        }
...
        tracing::info!("Waiting for timelines...");

```

And the set_stopping() call leads to connection_manager.rs::connection_manager_loop_step() exit because of:
```
...
    let mut timeline_state_updates = connection_manager_state
        .timeline
        .subscribe_for_state_updates();
...
                    match timeline_state_updates.changed().await {
                        Ok(()) => {
                            let new_state = connection_manager_state.timeline.current_state();
                            match new_state {
...
                                TimelineState::Broken { .. } | TimelineState::Stopping => {
                                    debug!("timeline entered terminal state {new_state:?}, stopping wal connection manager loop");
                                    return ControlFlow::Break(());
                                }
...
                ControlFlow::Break(()) => {
                    debug!("Timeline is no longer active, stopping wal connection manager loop");
                    return Err(Cancelled);
                }
...
```

The code corresponding to the call stack above:
```
   0: pageserver::tenant::timeline::walreceiver::walreceiver_connection::handle_walreceiver_connection::{{closure}}::{{closure}}
             at /__w/neon/neon/pageserver/src/tenant/timeline/walreceiver/walreceiver_connection.rs:449:33
```
is:
```
                    walingest   
                        .ingest_record(interpreted, &mut modification, &ctx)
                        .await
                        .with_context(|| {
                            format!("could not ingest record at {local_next_record_lsn}")
                        })
                        .inspect_err(|err| {
                            // TODO: we can't differentiate cancellation errors with
                            // anyhow::Error, so just ignore it if we're cancelled.
                            if !cancellation.is_cancelled() {
448:                                critical!("{err:?}")
                            }
                        })?;  
```

That is, the critical error was produced in that case, because cancellation.is_cancelled() was false at the moment. But this cancellation signaled later, inside timeline.shutdown(), walreceiver.cancel()...

With the following modification applied:
```
--- a/pageserver/src/tenant.rs
+++ b/pageserver/src/tenant.rs
@@ -3432,6 +3432,7 @@ async fn shutdown(
                 return Err(other);
             }
         };
+std::thread::sleep(core::time::Duration::from_millis(500));
 
         let mut js = tokio::task::JoinSet::new();
         {
```
I could reproduce this failure easily with:
DEFAULT_PG_VERSION=17 BUILD_TYPE=release poetry run pytest test_runner/regress/test_pageserver_restarts_under_workload.py::test_pageserver_restarts_under_worload -n 10 --count 10

It failed for me on iterations 3, 1, 19.

--- a/pageserver/src/tenant/timeline/walreceiver/walreceiver_connection.rs
+++ b/pageserver/src/tenant/timeline/walreceiver/walreceiver_connection.rs
@@ -445,7 +445,7 @@ pub(super) async fn handle_walreceiver_connection(
                         .inspect_err(|err| {
                             // TODO: we can't differentiate cancellation errors with
                             // anyhow::Error, so just ignore it if we're cancelled.
-                            if !cancellation.is_cancelled() {
+                            if !cancellation.is_cancelled() && !timeline.is_stopping() {
                                 critical!("{err:?}")
                             }
                         })?;
@@ -577,7 +577,7 @@ pub(super) async fn handle_walreceiver_connection(
                             .inspect_err(|err| {
                                 // TODO: we can't differentiate cancellation errors with
                                 // anyhow::Error, so just ignore it if we're cancelled.
-                                if !cancellation.is_cancelled() {
+                                if !cancellation.is_cancelled() && !timeline.is_stopping() {
                                     critical!("{err:?}")
                                 }
                             })?;

